crates/ide-db/src/path_transform.rs

```diff
@@ -298,9 +307,14 @@ fn get_syntactic_substs(impl_def: ast::Impl) -> Option<AstSubsts> {
 fn get_type_args_from_arg_list(generic_arg_list: ast::GenericArgList) -> Option<AstSubsts> {
     let mut result = AstSubsts::default();
     generic_arg_list.generic_args().for_each(|generic_arg| match generic_arg {
-        ast::GenericArg::TypeArg(type_arg) => result.types.push(type_arg),
+        // Const params are marked as consts on definition only,
+        // being passed to the trait they are indistguishable from type params;
+        // anyway, we don't really need to distinguish them here.
+        ast::GenericArg::TypeArg(type_or_const_arg) => {
+            result.types_and_consts.push(type_or_const_arg)
+        }
         ast::GenericArg::LifetimeArg(l_arg) => result.lifetimes.push(l_arg),
-        _ => (), // FIXME: don't filter out const params
+        _ => (),
```

Comment:
well, I've almost fixed it, one problem remains: `{0 + 1}` can cause syntax errors after substitution (e.g. `N * 5` -> `{0 + 1} * 5`). I've tried to simply parenthesize such expressions, but it just caused abother syntax error (`foo::<N>` -> `foo::<({0})>`). 

I feel like just leaving this error be for now, as it's a minor problem with no (?) simple solution

Reply:
Not sure if I follow; Rust has block expressions, so `{0 + 1} * 5` *is* a valid expression albeit in unusual style I believe?

Reply:
looks like it causes syntax errors only on older versions (which makes the problem even less serious) 

I've received an error [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021) for this snippet:
```
fn foo() -> usize{
    {1 + 1} * 5
}
```

Reply:
Ah okay, so when a statement starts with a block, there's an ambiguity:

> This can cause an ambiguity between it being parsed as a standalone statement and as a part of another expression; in this case, it is parsed as a statement. ([reference])

[reference]: https://doc.rust-lang.org/reference/statements.html#expression-statements

I agree with you that it's somewhat minor problem and we can leave it for now. Sorry for the confusion!